### DISCUSSION TITLE
Production Build

### DISCUSSION THREAD

**Discussion by jaybhattcscgfm** [shinsenter/php#51] [Q&A]
Hello,

I want to express my gratitude for your dedication to this repository. Really awesome work!

Our team utilizes the Docker image for development purposes. 

Firstly, I'm curious if there's a specific build tag intended for production usage. Despite the frequent changes, is there a stable tag or version we can rely on for building? It would be helpful if you could provide guidance on its usage in a production environment.

Additionally, are there any optimizations recommended for the production environment?

Thank you.

**Comment by shinsenter** [comment]
Hello, @jaybhattcscgfm 
Thank you for using and giving good feedback on my docker images.

Let me address your questions one by one.

> Despite the frequent changes, is there a stable tag or version we can rely on for building?

Regarding this question, I do not plan to use fixed tags (including the build date or commit id etc) for my docker images.

In order for the docker images to receive the latest updates not only of PHP but also of the base OS, Composer etc from upstreams, I have scheduled to build the docker images daily. Tagging fixed name for each build causes the number of inactive tags to increase significantly, making it difficult to search for tags. In addition, it also confuses users because we do not know the differences between two close builds, sometimes it is just a security patch in the base OS. Moreover, many projects tend to customize my docker images and create a base docker image for their projects, so a copy of my docker images is also kept through building that base image.

> It would be helpful if you could provide guidance on its usage in a production environment.

When you need to use one of my docker images for your project, you may pull a image tag, test it in your development environment, then retag it with a different name and push it to your private docker repository. That is one of the solutions I can suggest.

For example:
```shell
docker pull shinsenter/php:8.3-fpm-apache
docker tag shinsenter/php:8.3-fpm-apache your-private-repo/php:8.3-fpm-apache-20240208
docker push your-private-repo/php:8.3-fpm-apache-20240208
```
As in this example, you can rely on the docker image `your-private-repo/php:8.3-fpm-apache-20240208` (which is a copy of my docker image on the Feb 8th 2024) for your production and development environments for building.


> Additionally, are there any optimizations recommended for the production environment?

Basically you can consider that my docker images are ready for production purposes. These docker images are currently using the settings in the `php.ini-production` file that PHP recommends for production environments:
https://github.com/shinsenter/php/blob/1310d7c13b8d492a9cc1bdb1c5ed70b3aeabb637/src/php/common/php-ini-directives.dockerfile#L18-L21

However, depending on each project, fine tuning some PHP directives is necessary (e.g. max_execution_time, memory_limit etc), and I think your team just needs to change some directives to suit the production as well as development environments of your project. (Learn more here: https://www.php.net/manual/en/ini.core.php)

If you have any other questions, please feel free to ask. Or if you have any ideas that could improve my docker images, please let me know.

Best regards

**Comment by shinsenter** [comment]
@jaybhattcscgfm 

For stable versions you can depend on in production, we will apply version tags, but on another repo. https://hub.docker.com/r/shinsenter/php-archives/tags

This way, I think you can get frequently updated images under static tags, providing both latest code and production stability.

Best regards
